### PR TITLE
fix: allow services to indicate their dependencies

### PR DESCRIPTION
**Pull request by achingbrain**
Some services depend on others, for example circuit relay v2 and kad-dht don't really work without identify so we need a way to prompt the user to configure these dependencies.

The change is to allow extending the components a service requires with a service map of the dependencies.

The components type in the `Libp2pInit` type is extended by the `ServiceMap` generic type so when the user configures a node without the required dependencies, TypeScript will fail to compile as the components being passed to the factory function don't have type overlap with the required components.

### CODE REVIEW

File: test/circuit-relay/discovery.node.ts [hunk 1]
```diff
@@ -17,6 +18,7 @@ describe('circuit-relay discovery', () => {
 
   beforeEach(async () => {
     // create relay first so it has time to advertise itself via content routing
+    // @ts-expect-error TODO: typescript cannot infer the service map type
```
Comment (achingbrain) [inline test/circuit-relay/discovery.node.ts:21]: I am not sure why TypeScript can't infer this, it seems to work elsewhere.

**Comment by joeltg** [comment]
Super cool!

This does feel like it gets close to the line where TypeScript stops being fun anymore... once you start playing the generics game it's hard to stop. Personally I love going crazy with types but it can start to slow people down; it's up to you to weigh those tradeoffs.

Just a thought: if the idea is that "we have some components that get passed around, which always include certain built-in components, ", then would changing everything to take a `Components extends BuiltInComponents` parameter make more sense than passing around the "service map" just to pass it back into `Components<ServiceMap> = BuiltInComponents & ServiceMap`? Might be a lot of refactoring but it could eliminate the extra "service map" type concept.

**Comment by maschad** [comment]
>  then would changing everything to take a Components extends BuiltInComponents parameter make more sense than passing around the "service map" just to pass it back into Components<ServiceMap> = BuiltInComponents & ServiceMap? Might be a lot of refactoring but it could eliminate the extra "service map" type concept.

That's a good point @joeltg , the original impetus to even have these _optional_  services was to reduce Libp2p surface area in consumers and also to [improve extendability through custom modules](https://github.com/libp2p/js-libp2p/pull/1563). I think at the time we went with the `ServiceMap`  because we were unsure how could have an interface that could facilitate them being accessed from a `services` key.  

I think having the [`DefaultComponents` extend the `BuiltInComponents`](https://github.com/libp2p/js-libp2p/pull/1762/files#diff-af796a020e973df5f3ef424841c9f55fdb45d09ffac0ca4715fc0692f257b2e3R35-R125) and also having that [configuration validated](https://github.com/libp2p/js-libp2p/issues/1573) would solve this issue without a large refactor.

> Edit: I guess the other question hanging over this is that the services refactor kinda suggests that people can name their services anything they want now, but in reality dependent services still rely on specific names. Maybe make sure this is communicated in the docs somewhere?

Whilst we do now support custom modules I agree the docs should reflect those limitation, perhaps we could have a section in the configuration doc where we indicate that some services require others [e.g. setup with relay requires identify](https://github.com/libp2p/js-libp2p/blob/master/doc/CONFIGURATION.md#setup-with-relay)

**Comment by achingbrain** [comment]
Discussion: https://github.com/libp2p/js-libp2p/discussions/2135

**Comment by achingbrain** [comment]
I'm going to close this in favour of the solution proposed in https://github.com/libp2p/js-libp2p/discussions/2135

**Pull request closed without merging**